
# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1749288611 0
# Node ID 0e35d7cc243f8482a182f19212f3299916a78f33
# Parent  245b096dc5499a986a311c59ed36b8fff4d42455
Bug 617287 - Ensure the correct form of 'news:' URIs without server. r=aleca


'[s]news:' URIs independent of a specific server have three slashes added to be
handled correctly internally. This restores the original form as defined in
RFC 5538 when used by functions such as `nsDocumentViewer::CopyLinkLocation()`
or `XULBrowserWindow.setOverLink()`.

Differential Revision: https://phabricator.services.mozilla.com/D252833

diff --git a/mailnews/base/src/nsMsgMailNewsUrl.cpp b/mailnews/base/src/nsMsgMailNewsUrl.cpp
--- a/mailnews/base/src/nsMsgMailNewsUrl.cpp
+++ b/mailnews/base/src/nsMsgMailNewsUrl.cpp
@@ -454,18 +454,36 @@ NS_IMETHODIMP nsMsgMailNewsUrl::GetSearc
 ////////////////////////////////////////////////////////////////////////////////////
 // End nsIMsgMailNewsUrl specific support
 ////////////////////////////////////////////////////////////////////////////////////
 
 ////////////////////////////////////////////////////////////////////////////////////
 // Begin nsIURI support
 ////////////////////////////////////////////////////////////////////////////////////
 
+// For '[s]news:' URIs independent of a specific server,
+// nsNntpUrl::SetSpecInternal had to add three slashes to have them parsed
+// correctly by nsStandardURL. This restores the correct format, see also
+// https://datatracker.ietf.org/doc/html/rfc5538#section-4.
+void unmungeNewsURL(nsACString& aSpec) {
+  if (aSpec.Length() < 8) {
+    return;
+  }
+  int32_t colon = aSpec.Find(":");
+  if (Substring(aSpec, colon - 4, 8).EqualsLiteral("news:///")) {
+    aSpec.Cut(colon + 1, 3);
+  }
+}
+
 NS_IMETHODIMP nsMsgMailNewsUrl::GetSpec(nsACString& aSpec) {
-  return m_baseURL->GetSpec(aSpec);
+  nsresult rv = m_baseURL->GetSpec(aSpec);
+  NS_ENSURE_SUCCESS(rv, rv);
+
+  unmungeNewsURL(aSpec);
+  return NS_OK;
 }
 
 nsresult nsMsgMailNewsUrl::CreateURL(const nsACString& aSpec, nsIURL** aURL) {
   nsCOMPtr<nsIURL> url;
   nsresult rv = NS_MutateURI(NS_STANDARDURLMUTATOR_CONTRACTID)
                     .SetSpec(aSpec)
                     .Finalize(url);
   NS_ENSURE_SUCCESS(rv, rv);
@@ -624,17 +642,21 @@ NS_IMETHODIMP nsMsgMailNewsUrl::EqualsEx
 
 NS_IMETHODIMP
 nsMsgMailNewsUrl::GetSpecIgnoringRef(nsACString& result) {
   return m_baseURL->GetSpecIgnoringRef(result);
 }
 
 NS_IMETHODIMP
 nsMsgMailNewsUrl::GetDisplaySpec(nsACString& aUnicodeSpec) {
-  return m_baseURL->GetDisplaySpec(aUnicodeSpec);
+  nsresult rv = m_baseURL->GetDisplaySpec(aUnicodeSpec);
+  NS_ENSURE_SUCCESS(rv, rv);
+
+  unmungeNewsURL(aUnicodeSpec);
+  return NS_OK;
 }
 
 NS_IMETHODIMP
 nsMsgMailNewsUrl::GetDisplayHostPort(nsACString& aHostPort) {
   return m_baseURL->GetDisplayHostPort(aHostPort);
 }
 
 NS_IMETHODIMP
diff --git a/mailnews/news/test/unit/test_uriParser.js b/mailnews/news/test/unit/test_uriParser.js
--- a/mailnews/news/test/unit/test_uriParser.js
+++ b/mailnews/news/test/unit/test_uriParser.js
@@ -100,25 +100,29 @@ var tests = [
   // No-authority uris
   {
     uri: "news:rec.games.pinball",
     server: null,
     folder: null,
     newsAction: Ci.nsINntpUrl.ActionGetNewNews,
     group: "rec.games.pinball",
     host: "",
+    spec: "news:rec.games.pinball",
+    displaySpec: "news:rec.games.pinball",
   },
   {
     uri: "news:message-id@some-host.invalid",
     server: null,
     folder: null,
     newsAction: Ci.nsINntpUrl.ActionFetchArticle,
     messageID: "message-id@some-host.invalid",
     group: "",
     key: 0xffffffff,
+    spec: "news:message-id@some-host.invalid",
+    displaySpec: "news:message-id@some-host.invalid",
   },
 
   // news-message://host/group#key
   {
     uri: "news-message://localhost/test.simple.subscribe#1",
     newsAction: Ci.nsINntpUrl.ActionFetchArticle,
     group: "test.simple.subscribe",
     key: 1,

